api/src/controllers/reviews.ts

```diff
@@ -3,185 +3,162 @@
-        {
-          $match: query,
-        },
-        {
-          $addFields: {
-            _id: {
-              $toString: '$_id',
-            },
-          },
-        },
-        {
-          $lookup: {
-            from: 'votes',
-            let: {
-              cmpUserID: ctx.session.passport?.user.id,
-              cmpReviewID: '$_id',
-            },
-            pipeline: [
-              {
-                $match: {
-                  $expr: {
-                    $and: [
-                      {
-                        $eq: ['$$cmpUserID', '$userID'],
-                      },
-                      {
-                        $eq: ['$$cmpReviewID', '$reviewID'],
-                      },
-                    ],
-                  },
-                },
-              },
-            ],
-            as: 'userVote',
-          },
-        },
-        {
-          $addFields: {
-            userVote: {
-              $cond: {
-                if: {
-                  $ne: ['$userVote', []],
-                },
-                then: {
-                  $getField: {
-                    field: 'score',
-                    input: {
-                      $first: '$userVote',
-                    },
-                  },
-                },
-                else: 0,
-              },
-            },
-          },
-        },
-      ];
-
-      const reviews = await Review.aggregate<ReviewData>(pipeline);
-      if (reviews) {
-        return reviews;
-      } else {
-        return [];
-      }
+      return await getReviews({ ...input }, ctx.session.userId);
     }),
 
   /**
    * Add a review
    */
   add: userProcedure.input(reviewSubmission).mutation(async ({ input, ctx }) => {
+    const userId = ctx.session.userId!;
     // check if user is trusted
-    const verifiedCount = await Review.find({
-      userID: ctx.session.passport!.user.id,
-      verified: true,
-    })
-      .countDocuments()
-      .exec();
-
-    const review = {
+    const { verifiedCount } = (
+      await db
+        .select({ verifiedCount: count() })
+        .from(review)
+        .where(and(eq(review.userId, userId), eq(review.verified, true)))
+    )[0];
+    const reviewToAdd = {
       ...input,
-      userDisplay: input.userDisplay === 'Anonymous Peter' ? 'Anonymous Peter' : ctx.session.passport!.user.name,
-      userID: ctx.session.passport!.user.id,
+      userId: userId,
       verified: verifiedCount >= 3, // auto-verify if use has 3+ verified reviews
     };
 
-    //check if review already exists for same professor, course, and user
-    const query: ReviewFilter = {
-      courseID: input.courseID,
-      professorID: input.professorID,
-      userID: ctx.session.passport?.user.id,
-    };
-
-    const reviews = await Review.find(query);
-    if (reviews?.length > 0)
+    //check if review already exists for same professor, course, and user (do this before verifying captcha)
+    const existingReview = await db
```

Comment:
Ideally this check happens before they even open the form – let's create an issue for that

Reply:
Ah, so you mean like doing the check to prevent them from even having the option on the form to create an identical review?

Reply:
#500